google/cloud/spark_connect/session.py

```diff
@@ -487,6 +489,51 @@ def _remove_stoped_session_from_file(self):
                     f"Exception while removing active session in file {file_path} , {e}"
                 )
 
+    """
+    Install PyPi packages (with their dependencies) in the active Spark 
+    session on the driver and executors. 
+    We recommend to pin the version to have consistent environment.
+    
+    Examples
+    --------
+    >>> spark.addArtifacts("pypi://spacy==3.8.4")
+    >>> spark.addArtifacts("pypi://spacy") 
+    
+    Notes
+    -----
+    Popular packages are already pre-installed in s8s runtime. 
+     https://cloud.google.com/dataproc-serverless/docs/concepts/versions/spark-runtime-2.2#python_libraries
+    This is an API available only in Google Spark Session as of today.
+    If there are conflicts/package doesn't exist, it throws an exception.
+    """
+
+    def addArtifact(self, package: str) -> None:
+        if package in self._installed_pypi_libs:
+            logger.info("Ignoring as artifact has already been added earlier")
+            return
+
+        if package.startswith("pypi://") is False:
+            raise ValueError(
+                "Only PyPi packages are supported in format `pypi://spacy`"
+            )
+
+        dependencies = {"Version": "1.0", "packages": [package]}
```

Comment:
Is the `Version` a version of the dependency file? If yes then shouldn't it be incremented each time `GoogleSparkSession.addArtifact` is called ?

Reply:
Yeah, the semantics of this version package is unclear to me. What is the intent here? Also, the use of a package list inside almost suggests some atomic package set, so maybe my above comment doesn't actually apply.

Reply:
The "Version" is for the file format we are introducing to be future proof in case we add support for Conda etc or decide to serialize list of dependencies in some other format.
On server side, we will add handler based on the version.

Regarding installation, each time a new file (currently containing only 1 package) will be created and "addArtifacts" will be called.

Reply:
Changed to "client-version" which will be tied to GoogleSparkSession.